dymos/transcriptions/transcription_base.py

```diff
@@ -286,157 +287,107 @@ def setup_timeseries_outputs(self, phase):
-            con_options['linear'] = linear
-
-            bc_comp._add_constraint(con_name, **con_options)
-
-        if bc_comp:
-            bc_comp.configure_io()
-
-        for var, options in bc_dict.items():
-            con_name = options['constraint_name']
-
-            src, shape, units, linear = self._get_boundary_constraint_src(var, loc, phase)
-
-            size = np.prod(shape)
-
-            # Build the correct src_indices regardless of shape
-            if loc == 'initial':
-                src_idxs = np.arange(size, dtype=int).reshape(shape)
+        Returns
+        -------
+        con_output : str
+            The phase-relative path being constrained.
+        constraint_kwargs : dict
+            Keyword arguments for the OpenMDAO add_constraint method.
+        """
+        num_nodes = self._get_num_timeseries_nodes()
+
+        constraint_kwargs = {key: options for key, options in options.items()}
+        con_name = constraint_kwargs.pop('constraint_name')
+
+        # Determine the path to the variable which we will be constraining
+        var = options['name']
+        var_type = phase.classify_var(var)
+
+        # These are the flat indices at a single point in time used
+        # in either initial, final, or path constraints.
+        idxs_in_initial = phase._indices_in_constraints(var, 'initial')
+        idxs_in_final = phase._indices_in_constraints(var, 'final')
+        idxs_in_path = phase._indices_in_constraints(var, 'path')
+
+        size = np.prod(options['shape'], dtype=int)
+
+        flat_idxs = get_constraint_flat_idxs(options)
+
+        # Now we need to convert the indices given by the user at any given point
+        # to flat indices to be given to OpenMDAO as flat indices spanning the phase.
+        if var_type == 'parameter':
+            if any([idxs_in_initial.intersection(idxs_in_final),
+                    idxs_in_initial.intersection(idxs_in_path),
+                    idxs_in_final.intersection(idxs_in_path)]):
+                raise RuntimeError(f'In phase {phase.pathname}, parameter `{var}` is subject to multiple boundary '
+                                   f'or path constraints.\nParameters are single values that do not change in '
+                                   f'time, and may only be used in a single boundary or path constraint.')
+            constraint_kwargs['indices'] = flat_idxs
+        else:
+            if constraint_type == 'initial':
+                constraint_kwargs['indices'] = flat_idxs
+            elif constraint_type == 'final':
+                constraint_kwargs['indices'] = size * (num_nodes - 1) + flat_idxs
             else:
-                src_idxs = np.arange(-size, 0, dtype=int).reshape(shape)
+                # This is a path constraint.
+                # Remove any flat indices involved in an initial constraint from the path constraint
+                idxs_not_in_initial = list(set(flat_idxs.tolist()) - idxs_in_initial)
```

Comment:
Not sure how big flat_idxs can get, but you compute `list(set(flat_idxs.tolist())` twice.  Also if flat_idxs is already a flat array you probably don't need the `.tolist()` either.